--- orig/config.php	2022-06-05 11:16:36.000000000 +0200
+++ docker/config.php	2022-06-06 11:46:31.000000000 +0200
@@ -319,14 +319,16 @@
    'ldaps://ldap.example.com/',
    'ldapi://%2fusr%local%2fvar%2frun%2fldapi'
            (Unix socket at /usr/local/var/run/ldap) */
-$servers->setValue('server','host','127.0.0.1');
+//$servers->setValue('server','host','127.0.0.1');
+$servers->setValue('server','host',  getenv("PLA_HOST" , true) ?: '127.0.0.1');
 
 /* The port your LDAP server listens on (no quotes). 389 is standard. */
 // $servers->setValue('server','port',389);
 
 /* Array of base DNs of your LDAP server. Leave this blank to have phpLDAPadmin
    auto-detect it for you. */
-$servers->setValue('server','base',array('dc=example,dc=com'));
+//$servers->setValue('server','base',array('dc=example,dc=com'));
+$servers->setValue('login','bind_id',  getenv("PLA_ADMIN" , true) ?: 'cn=admin,dc=example,dc=com');
 
 /* Five options for auth_type:
    1. 'cookie': you will login via a web form, and a client-side cookie will
@@ -353,8 +355,9 @@
    auth_type, then you can also specify the bind_id/bind_pass here for searching
    the directory for users (ie, if your LDAP server does not allow anonymous
    binds. */
-$servers->setValue('login','bind_id','cn=admin,dc=example,dc=com');
+//$servers->setValue('login','bind_id','cn=admin,dc=example,dc=com');
 #  $servers->setValue('login','bind_id','cn=Manager,dc=example,dc=com');
+$servers->setValue('login','bind_id',  getenv("PLA_ADMIN" , true) ?: 'cn=admin,dc=example,dc=com');
 
 /* Your LDAP password. If you specified an empty bind_id above, this MUST also
    be blank. */
@@ -573,6 +576,7 @@
    'unique','dn' and 'unique','pass' option will not let updates to
    occur with other attributes have the same value. */
 #  $servers->setValue('unique','attrs',array('mail','uid','uidNumber'));
+$servers->setValue('unique','attrs', explode(' ', getenv("PLA_UNIQUE_ATTRS" , true) ?:'mail uid uidNumber'));
 
 /* If you set this, then phpldapadmin will bind to LDAP with this user ID when
    searching for attribute uniqueness. The idea is, this user id would have full
